rx_signal_processing/decimate.cu

```diff
@@ -1,221 +1,693 @@
+ * @param[in]  num_taps_per_filter  Number of taps per filter.
+ * @param[in]  num_freqs            Number of receive frequencies.
+ * @param[in]  num_antennas         Number of antennas for which there are samples.
+ * @param[in]  stream               CUDA stream with which to associate the invocation of the
+ *                                  kernel.
+ */
+void bandpass_decimate2048_wrapper(cuComplex* original_samples,
+  cuComplex* decimated_samples,
+  cuComplex* filter_taps, uint32_t dm_rate,
+  uint32_t samples_per_antenna, uint32_t num_taps_per_filter, uint32_t num_freqs,
+  uint32_t num_antennas, cudaStream_t stream) {
+
+  //Allocate shared memory on device for all filter taps.
+  auto shr_mem_taps = num_freqs * num_taps_per_filter * sizeof(cuComplex);
+  std::cout << "    Number of shared memory bytes: "<< shr_mem_taps << std::endl;
+
+  auto dimGrid = create_bandpass_grid(samples_per_antenna, dm_rate, num_antennas);
+  auto dimBlock = create_bandpass_block(num_taps_per_filter/2, num_freqs);
+  bandpass_decimate2048<<<dimGrid,dimBlock,shr_mem_taps,stream>>>(original_samples, decimated_samples,
+    filter_taps, dm_rate, samples_per_antenna);
 }
 
-static dim3 create_grid(uint32_t num_samples, uint32_t dm_rate, uint32_t num_channels) // REVIEW #26 no more channels
-{
-    auto num_blocks_x = num_samples/dm_rate;
-    auto num_blocks_y = num_channels;
-    auto num_blocks_z = 1;
-    std::cout << "    Grid size: " << num_blocks_x << " x " << num_blocks_y << " x "
-        << num_blocks_z << std::endl;
-    dim3 dimGrid(num_blocks_x,num_blocks_y,num_blocks_z);
-
-    return dimGrid;
+/**
+ * @brief      Performs decimation using a lowpass filter on one or more sets of baseband samples 
+ * corresponding to each RX frequency. This algorithm works on filters with less that 1024 taps.
+ *
+ * @param[in]  original_samples     A pointer to input samples for one or more baseband datasets.
+ * @param[in]  decimated_samples    A pointer to a buffer to place output samples for each frequency
+ *                                  dataset after decimation.
+ * @param[in]  filter_taps          A pointer to a lowpass filter used for further decimation.
+ * @param[in]  dm_rate              Decimation rate.
+ * @param[in]  samples_per_antenna  The number of samples per antenna in the original set of
+ *                                  samples.
+ * 
+ * This function performs a parallel version of filtering+downsampling on the GPU to be able
+ * process data in realtime. This algorithm will use 1 GPU thread per filter tap if there are less
+ * than 1024 taps for all filters combined. Only works with power of two length filters, or a
+ * filter that is zero padded to a power of two in length. This algorithm takes one or more 
+ * baseband datasets corresponding to each RX frequency and filters each one using a single lowpass
+ * filter before downsampling.
+ * 
+ *   gridDim.x - The number of decimated output samples for one frequency set.
```

Comment:
instead of the word 'set' can this be 'for one frequency and one antenna'? would be more clear